Added optional no_std feature

ParkerTenBroeck opened the pull request:
The AVec module is disabled when no_std feature is enabled.
all tests passed on nightly and stable with and without no_std enabled

ParkerTenBroeck committed:
added optional no_std feature

ParkerTenBroeck committed:
fixed test dependency not being present for tests on nightly

ParkerTenBroeck committed:
fixed test feature not being enabled for nightly no_std

notflan commented on src/lib.rs:
```diff
@@ -157,7 +167,15 @@ where F: FnOnce(&mut [MaybeUninit<u8>]) -> T
+
+        #[cfg(feature = "no_std")]
+	    {
+            rval = MaybeUninit::new(callback(slice));
```
From what I can see, the removal of `catch_unwind()` here may cause the following scenario:
 * If downstream crate `A` pulls in `stackalloc` with feature `no_std` (either via a direct or transitive dependency,) but itself is **not** `#![no_std]`, the `std` panic handler will be registered by `A`.
 * If `A` has not set `panic = "abort"` or registered its own panic handler somewhere that does not attempt an unwind.
 * If a closure passed to any `stackalloc` function panics (there's no reliable way to tell,) then we are not catching the unwind at the FFI boundary. (Of a C function trampoline with a variable-sized stack, too.)

This situation would be very unsound, and would make any use of `stackalloc` by `A` completely unsafe as we cannot bound these functions (at compile-time or runtime) to take only closures that we are sure will not be unwound through from an `std`-enabled panic without `catch_unwind()`. Sure, it is possible for a downstream crate to do something silly like, I don't know, allow a C++ FFI function to throw an exception that gets past the Rust FFI boundary, but FFI functions must be marked `unsafe` anyway so the programmer is aware this may happen.
This is a soundness problem for us though because `stackalloc` would silently allow undefined behaviour to happen here within a safe Rust interface.

ParkerTenBroeck replied:
It is possible to implement a custom `catch_unwind` in a `no_std` environment however there are some drawbacks. 

- aborting panics behave the same but unwinding panics within the closure will turn into aborting panics after being caught. This is because with no std present you must call onto OS-specific libraries to produce unwinding panics making it extremely hard to resume the unwinding process. 

- The same ordeal goes for the clean up of the panic 
`extern "C" {
    fn __rust_panic_cleanup(payload: *mut u8) -> *mut (dyn Any + Send + 'static);
}` is a core part of  `catch_unwind` and is not present without the std.

- The implementation of the custom `catch_unwind` directly calls `core::intrinsics::r#try` which is unstable and requires the use of nightly

With the custom `catch_unwind` any unwinding panic should be caught but unwinding will not be continued and an aborting panic will be produced. This is irreverent for most cases with no_std but in the rare cases that a unwinding panic is produced it will provide a safety net. The only cases where I can see this happening are  
- a downstream crate that enables the `no_std` feature for this crate while using std itself. 
- a downstream crate directly calls upon OS-specific unwinding functionality. In those cases aborting panics should also be produced.

> Once inline assembly is a stable feature I would like to replace the FFI trampoline with a #[naked] assembly-implemented dynamic stack allocation/deallocation function trampoline (or remove the primary trampoline entirely and just use inline asm! within the main function that calls into the closure) in pure Rust.
However, currently, having an FFI trampoline is the only way to implement dynamic stack growth in Rust on stable. Perhaps limiting safe no_std support to nightly only and doing this now would allow a fix of this problem but it would be unfortunate to make the no_std feature depend on a nightly compiler.

To my knowledge using the `no_std` feature requires the use of `lang_items` and `lang_items` requires the use of nightly so making the no_std feature only available on nightly seems like a non-issue


> If we should install our own panic handler when feature no_std is enabled to prevent an unwind by a downstream crate's use of std's default panic handler? I don't know if this would prevent downstream crates from installing their own panic handler entirely, I think it would but perhaps it's possible to install a panic handler per-crate, or have multiple panic handlers, or set our own default panic handler that can be overridden by a downstream crate that is #![no_std] itself (i.e. just prevent the std panic handler from being registered by the bottom-level crate of the dependency graph)? That's... probably not the case unfortunately but again I don't really know much about it...

You can only define a single panic handler per compiled program. That means if we add our own panic_handler it would prevent upstream crate from defining their own

ParkerTenBroeck replied:
> Once inline assembly is a stable feature 

Inline assembly was stabilized in 1.59.0 for the following architectures
- x86 and x86-64
- ARM
- AArch64
- RISC-V

notflan replied:
Oh, thank you. There was a lot I didn't know there.
Huh, I did not know even know inline assembly was already stable...

notflan reviewed (changes requested):
Hiya, thanks for adding this feature! 
I am fairly unfamiliar with `no_std` in general, as it is not something I have used myself. However, I notice the removal of `std::panic::catch_unwind()` when calling the user-provided closure since (I assume) catching unwinds this way is not available without `std`, as the stack-unwinding mechanism of a panic is implemented *by* `std`.

As I am not familiar with `no_std` support and use by dependants, I am not sure if this is actually a problem that may occur, could you offer insight on this?

There may be some explanations but all I can think of now is:
 * This might not even be a problem scenario at all for some reasons I don't know. (I might be missing something here, using `no_std` is not an area of Rust I am very familiar with. Maybe it's even disallowed pulling in a `#![no_std]` crate as a dependency for one using `std` in the first place, I'm not sure.)
 * If there may be another way to mitigate or prevent this problem from happening (since there is no way to ensure a function does not cause an unwind.)
 * If `std`'s unwind actually can, (somehow?) unwind through an FFI C trampoline function with a variably-sized stack-frame and that behaviour can be depended on. (I doubt this is true, but I don't know for sure.)
 * If `std`'s unwind behaviour is itself standardised (which I don't think it is, but, maybe?), I may be able to modify the FFI function responsible for calling this trampoline callback to have a cleanup section that would correctly reset the stack pointer. But this is a longshot as I am unfamiliar with how `std` actually unwinds the stack. It may not even be possible at all.
 * If we should install our own panic handler when feature `no_std` is enabled to prevent an unwind by a downstream crate's use of `std`'s default panic handler? I don't know if this would prevent downstream crates from installing their own panic handler entirely, I think it would but perhaps its possible to install a panic handler per-crate, or have multiple panic handlers, or set our own default panic handler that can be overriden by a downstream crate that is `#![no_std]` itself (i.e. just prevent the `std` panic handler from being registered by the bottom-level crate of the dependency graph)? That's... probably not the case unfortunately but again I don't really know much about it...
 * If we should just mark *every* `stackalloc` function as `unsafe` when the feature `no_std` in enabled and document that the installed panic handler of a dependant crate **must not** attempt unwind the stack past the point of calling one of these functions.
 * Once inline assembly is a stable feature I would like to replace the FFI trampoline with a `#[naked]` assembly-implemented dynamic stack allocation/deallocation function trampoline (or remove the primary trampoline entirely and just use inline `asm!` within the main function that calls into the closure) in pure Rust.
However, currently, having an FFI trampoline is the only way to implement dynamic stack growth in Rust on stable. Perhaps limiting safe `no_std` support to nightly only and doing this now would allow a fix of this problem but it would be unfortunate to make the `no_std` feature depend on a nightly compiler.

When I have more time I will test this myself and read more about how unwinds are implemented in `std`, but if you have any ideas (or if I'm just wrong >_<) please let me know.
For now I will leave this PR open because I'm not sure how safe it would be to merge as-is, but thank you for the work you put in doing this <3

ParkerTenBroeck committed:
Added catch_unwind for no_std

added catch_unwind functionality for no_std where unwinding panics are turned into aborting panics

notflan commented:
Okay, merging this now, from what you've said I don't see any issues here. Thanks for your work on this! 
I'll push a new update to crates.io to add the feature when I can get my laptop later today.
